src/satosa/satosa_config.py

```diff
@@ -44,26 +46,49 @@ def __init__(self, config):
 
         self._verify_dict(self._config)
 
+        self._load_plugins()
+        self._load_internal_attributes()
+
+    def _parse_config(self, config):
+        return next(
+            filter(
+                lambda conf: conf is not None,
+                map(
+                    lambda parser: parser(config),
+                    (self._load_dict, self._load_yaml),
+                ),
+            ),
+            None,
+        )
+
+    def _load_plugins(self):
+        def load_plugin_config(config):
+            plugin_config = self._parse_config(config)
+            if not plugin_config:
+                raise SATOSAConfigurationError(
+                    "Failed to load plugin config '{}'".format(config)
+                )
+            else:
+                return plugin_config
```

Comment:
We should try to have a straight codepath. `else` is not needed here.

Reply:
Thought that "explicit is better than implicit" for the eye here, but sure

Reply:
`if/else` says that at this point there one of two things that you can follow. This is not the case here. What we do here is just check and handle an error case. To make this clearer forget that the `else` part is a `return` statement. `return` is just a command, and it is only by chance that there only one command here. If you imagine we had 4 commands would you still nest them under the `else` branch?

```python
def foo(x):
    value = command1(x)
    if has_error(value):
        generate_error()
    else:
        command2(value)
        command3(value)
        command4(value)
```
or 
```python
def foo(x):
    value = command1(x)
    if has_error(value):
        generate_error()
    command2(value)
    command3(value)
    command4(value)
```

Reply:
It depends is the answer again. I would probably have a validation function call to throw the error and hide the `if` from `foo` so as to keep in main view one logical exit path

Reply:
> to keep in main view one logical exit path

I feel we're saying the same thing. To keep one logical path you would not put the rest of the code under the else branch.